### DISCUSSION TITLE
After upgrading to cypress 13 all my test fail with blank page after the first iteration

### DISCUSSION THREAD

**Discussion by easaborit** [cypress-io/cypress#28216] [Questions and Help]
After upgrading to Cypress 13, all tests failed with blank pages after the first IT

I was using Cypress 10.10.0, and I upgraded to the latest version. After this, all tests start failing by showing a blank page.

I tried to set the test isolation to true, 
I have this login: in session_command.js
```
module.exports = {
    saveLocalStorage: () => {
        Object.keys(localStorage).forEach(key => {
            LOCAL_STORAGE_MEMORY[key] = localStorage[key];
        });
    },
    restoreLocalStorage: () => {
        Object.keys(LOCAL_STORAGE_MEMORY).forEach(key => {
            localStorage.setItem(key, LOCAL_STORAGE_MEMORY[key]);
        });
    },
    login: (username, password, saveLocalStorage = true) => {
        cy.session([username, password], () => {
            cy.start();
            cy.get('#signin').type(username ?? Cypress.env('auth_username'));
            cy.get('#signin-password').type(password ?? Cypress.env('auth_password'));
            cy.get('#signin-submit').click();
            cy.get('div.Organizations').should('be.visible');
        });
        saveLocalStorage && cy.saveLocalStorage();
    },
}
```

and also I have an example test:

```
describe( 'Contest page', () => {
    before(() => {
        cy.login();
    });

    beforeEach(() => {
        cy.restoreLocalStorage();
    });

        it('Check and validate name', ()  => {
           // do something
          
            });
        });
```

**Comment by portyu9** [comment]
Hi @easaborit 

The blank page is expected with `testIsolation: true`.

With test isolation enabled, Cypress clears the page by visiting `about:blank` before every test. `cy.session()` also clears/restores the browser context, and after `cy.session()` finishes you need to explicitly visit the page you want to test.

Your current structure logs in only once:

```js
before(() => {
  cy.login()
})
```

but every subsequent `it()` starts with a clean page.

Use `cy.session()` in `beforeEach()` and visit the page afterwards:

```js
describe('Contest page', () => {
  beforeEach(() => {
    cy.login()
    cy.visit('/contest')
  })

  it('test 1', () => {
    // ...
  })

  it('test 2', () => {
    // ...
  })
})
```

Your login command can cache the expensive login:

```js
Cypress.Commands.add('login', (username, password) => {
  cy.session([username, password], () => {
    cy.visit('/login')

    cy.get('#signin').type(username)
    cy.get('#password').type(password)
    cy.get('button[type="submit"]').click()

    cy.url().should('not.include', '/login')
  }, {
    validate() {
      // Use an authenticated request or another reliable check here.
    },
  })
})
```

Then:

```js
beforeEach(() => {
  cy.login(username, password)
  cy.visit('/contest')
})
```

You should not need the manual `saveLocalStorage()` / `restoreLocalStorage()` workaround for authentication state. `cy.session()` already caches and restores cookies, `localStorage`, and `sessionStorage`.

You *can* disable test isolation for a suite, but that makes tests depend on state left behind by earlier tests. The recommended approach is to keep isolation enabled, restore the session in `beforeEach()`, and explicitly `cy.visit()` the page each test requires.

Please mark this answer as accepted if it helped, thank you.
